compiler/damlc/daml-lf-conversion/src/DA/Daml/LFConversion.hs

```diff
@@ -1233,32 +1228,93 @@ convertExpr env0 e = do
+                    resultType' <- convertType env resultType
+                    pure $ mkCase env ty resultType' bind' [alt']
 
       where
         asLet = convertLet env bind scrutinee $ \env -> convertExpr env x
-    go env (Case scrutinee bind typ []) args = fmap (, args) $ do
-        -- GHC only generates empty case alternatives if it is sure the scrutinee will fail, LF doesn't support empty alternatives
-        scrutinee' <- convertExpr env scrutinee
-        typ' <- convertType env typ
-        bind' <- convVarWithType env bind
-        pure $
-          ELet (Binding bind' scrutinee') $
-          ECase (EVar $ convVar bind) [CaseAlternative CPDefault $ EBuiltin BEError `ETyApp` typ' `ETmApp` EBuiltin (BEText "Unreachable")]
-    go env (Case scrutinee bind _ (defaultLast -> alts)) args = fmap (, args) $ do
+    go env (Case scrutinee bind resultType (defaultLast -> alts)) args = fmap (, args) $ do
         scrutinee' <- convertExpr env scrutinee
         bindTy <- convertType env $ varType bind
         alts' <- mapM (convertAlt env bindTy) alts
         bind' <- convVarWithType env bind
-        if isDeadOcc (occInfo (idInfo bind))
-          then pure $ ECase scrutinee' alts'
-          else pure $
+        resultType' <- convertType env resultType
+        if isDeadOcc (occInfo (idInfo bind)) && all isNormalCaseAlternative alts'
+        then pure $ mkCase env bindTy resultType' scrutinee' alts'
+        else pure $
             ELet (Binding bind' scrutinee') $
-            ECase (EVar $ convVar bind) alts'
+            mkCase env bindTy resultType' (EVar $ convVar bind) alts'
     go env (Let (Rec xs) _) args = unsupported "Local variables defined recursively - recursion can only happen at the top level" $ map fst xs
     go env o@(Coercion _) args = unhandled "Coercion" o
     go _ x args = unhandled "Expression" x
 
+-- | Represents a case alternative with a generalized case pattern.
+data GeneralisedCaseAlternative
+    = GCAEquality EqualityAlternative
+        -- ^ Pattern matching via built-in equality.
+    | GCANormal CaseAlternative
+        -- ^ Normal case alternative that is directly supported by LF.
+        -- This includes the default case (CPDefault).
+    deriving (Eq, Ord)
+
+-- | Is this a normal case alternative?
+isNormalCaseAlternative :: GeneralisedCaseAlternative -> Bool
+isNormalCaseAlternative = \case
+    GCANormal _ -> True
+    _ -> False
+
+-- | Represents a case alternative that is not directly supported
+-- by LF's CaseAlternative syntax. In particular the pattern here
+-- is some expression that we want to use the built-in equality
+-- to match.
+data EqualityAlternative = EqualityAlternative
+    { eqaltPattern :: LF.Expr
+        -- ^ Pattern expression to test against. This is an inert
+        -- expression (can't raise errors, e.g. a constructor).
+    , eqaltBody :: LF.Expr
+        -- ^ Right-hand side of case alternative. Potentially divergent.
+    } deriving (Eq, Ord)
```

Comment:
Just an idea:
Currently you factor this into
equality = eqExpr * rhs
other = pattern * rhs

Would it be a good idea to factor it into?
alternative = pattern * rhs
pattern = casePattern | equality

Reply:
That's what I had at first and it was slightly more annoying to work with. I think I'll keep it like this for now and we can switch it later if it becomes more helpful.

Reply:
I went ahead and factored it as you suggested :-)